Issue while building on OS X  : "error: constexpr constructor never produces a constant expression"

I am building Caliper with Spack and see following : 

```
$ spack install -j 1 -v caliper ^python@2.7
....
→ spack install -v caliper ^python@2.7
==> cmake@3.5.2 : externally installed in /usr/local-fake
==> cmake@3.5.2 : already registered in DB
==> mpich@3.2 : externally installed in /usr/local/Cellar/open-mpi/4.0.0
==> mpich@3.2 : already registered in DB
==> python@2.7.10 : externally installed in /System/Library/Frameworks/Python.framework/Versions/2.7
==> python@2.7.10 : already registered in DB
==> Installing caliper
==> Searching for binary cache of caliper
==> Warning: No Spack mirrors are currently configured
==> No binary for caliper found: installing from source
==> Using cached archive: /Users/kumbhar/workarena/software/sources/spack/var/spack/cache/caliper/caliper-2.0.1.tar.gz
==> Warning: Fetching from mirror without a checksum!
  This package is normally checked out from a version control system, but it has been archived on a spack mirror.  This means we cannot know a checksum for the tarball in advance. Be sure that your connection to this mirror is secure!
==> Staging archive: /Users/kumbhar/workarena/software/sources/spack/var/spack/stage/caliper-2.0.1-abbxhgovgy4z7s3lw36rrbd4uvejyzfv/caliper-2.0.1.tar.gz
==> Created stage in /Users/kumbhar/workarena/software/sources/spack/var/spack/stage/caliper-2.0.1-abbxhgovgy4z7s3lw36rrbd4uvejyzfv
==> No patches needed for caliper
==> Building caliper [CMakePackage]
==> Executing phase: 'cmake'
==> [2019-03-19-20:27:18.052209] 'cmake' '/Users/kumbhar/workarena/software/sources/spack/var/spack/stage/caliper-2.0.1-abbxhgovgy4z7s3lw36rrbd4uvejyzfv/Caliper' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:PATH=/Users/kumbhar/workarena/software/sources/spack/opt/spack/darwin-mojave-x86_64/clang-9.0.0-apple/caliper-2.0.1-abbxhgovgy4z7s3lw36rrbd4uvejyzfv' '-DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_FIND_FRAMEWORK:STRING=LAST' '-DCMAKE_FIND_APPBUNDLE:STRING=LAST' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=FALSE' '-DCMAKE_INSTALL_RPATH:STRING=/Users/kumbhar/workarena/software/sources/spack/opt/spack/darwin-mojave-x86_64/clang-9.0.0-apple/caliper-2.0.1-abbxhgovgy4z7s3lw36rrbd4uvejyzfv/lib;/Users/kumbhar/workarena/software/sources/spack/opt/spack/darwin-mojave-x86_64/clang-9.0.0-apple/caliper-2.0.1-abbxhgovgy4z7s3lw36rrbd4uvejyzfv/lib64;/usr/local/Cellar/open-mpi/4.0.0/lib' '-DCMAKE_PREFIX_PATH:STRING=/System/Library/Frameworks/Python.framework/Versions/2.7;/usr/local/Cellar/open-mpi/4.0.0;/usr/local-fake' '-DBUILD_TESTING=Off' '-DBUILD_DOCS=Off' '-DBUILD_SHARED_LIBS=On' '-DWITH_DYNINST=Off' '-DWITH_CALLPATH=Off' '-DWITH_GOTCHA=Off' '-DWITH_PAPI=Off' '-DWITH_LIBPFM=Off' '-DWITH_SOSFLOW=Off' '-DWITH_SAMPLER=Off' '-DWITH_MPI=On' '-DMPI_C_COMPILER=/usr/local/Cellar/open-mpi/4.0.0/bin/mpicc' '-DMPI_CXX_COMPILER=/usr/local/Cellar/open-mpi/4.0.0/bin/mpic++'
-- The C compiler identification is AppleClang 9.0.0.9000037
-- The CXX compiler identification is AppleClang 9.0.0.9000037
-- Check for working C compiler: /Users/kumbhar/workarena/software/sources/spack/lib/spack/env/clang/clang
-- Check for working C compiler: /Users/kumbhar/workarena/software/sources/spack/lib/spack/env/clang/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Users/kumbhar/workarena/software/sources/spack/lib/spack/env/clang/clang++
-- Check for working CXX compiler: /Users/kumbhar/workarena/software/sources/spack/lib/spack/env/clang/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE
-- Found MPI_C: /usr/local/Cellar/open-mpi/4.0.0/lib/libmpi.dylib (found version "3.1")
-- Found MPI_CXX: /usr/local/Cellar/open-mpi/4.0.0/lib/libmpi.dylib (found version "3.1")
-- Found MPI: TRUE (found version "3.1")
-- Found PythonInterp: /System/Library/Frameworks/Python.framework/Versions/2.7/bin/python (found version "2.7.10")
-- Caliper configuration summary:
-- Caliper version           : 2.0.1
-- Build type                : RelWithDebInfo
-- Compiler                  : AppleClang 9.0.0.9000037 (/Users/kumbhar/workarena/software/sources/spack/lib/spack/env/clang/clang++)
-- System                    : Darwin-18.0.0 (x86_64)
-- Install dir               : /Users/kumbhar/workarena/software/sources/spack/opt/spack/darwin-mojave-x86_64/clang-9.0.0-apple/caliper-2.0.1-abbxhgovgy4z7s3lw36rrbd4uvejyzfv
-- Build shared libs         : On
-- Build Caliper tools       : ON
-- GOTCHA support            : No
-- PAPI support              : No
-- Libpfm support            : No
-- Libunwind support         : No
-- Dyninst support           : No
-- Sampler support           : No
-- SOSFlow support           : No
-- MPI support               : Yes, using /usr/local/Cellar/open-mpi/4.0.0/lib/libmpi.dylib
-- MPIWRAP support           : Yes, using PMPI
-- MPIT support              : No
-- OMPT support              : No
-- NVProf support            : No
-- CUpti support             : No
-- TAU support               : No
-- VTune support             : No
-- Configuring done
-- Generating done
..........

[ 40%] Building CXX object src/services/CMakeFiles/caliper-services.dir/aggregate/Aggregate.cpp.o
cd /tmp/kumbhar/spack-stage/spack-stage-IWqgNK/Caliper/spack-build/src/services && /Users/kumbhar/workarena/software/sources/spack/lib/spack/env/clang/clang++   -I/tmp/kumbhar/spack-stage/spack-stage-IWqgNK/Caliper/spack-build/include -I/Users/kumbhar/workarena/software/sources/spack/var/spack/stage/caliper-2.0.1-abbxhgovgy4z7s3lw36rrbd4uvejyzfv/Caliper/include -I/tmp/kumbhar/spack-stage/spack-stage-IWqgNK/Caliper/spack-build -I/Users/kumbhar/workarena/software/sources/spack/var/spack/stage/caliper-2.0.1-abbxhgovgy4z7s3lw36rrbd4uvejyzfv/Caliper/src -I/tmp/kumbhar/spack-stage/spack-stage-IWqgNK/Caliper/spack-build/src/services  -O2 -g -DNDEBUG -fPIC   -std=gnu++11 -o CMakeFiles/caliper-services.dir/aggregate/Aggregate.cpp.o -c /Users/kumbhar/workarena/software/sources/spack/var/spack/stage/caliper-2.0.1-abbxhgovgy4z7s3lw36rrbd4uvejyzfv/Caliper/src/services/aggregate/Aggregate.cpp
In file included from /Users/kumbhar/workarena/software/sources/spack/var/spack/stage/caliper-2.0.1-abbxhgovgy4z7s3lw36rrbd4uvejyzfv/Caliper/src/services/aggregate/Aggregate.cpp:49:
/Users/kumbhar/workarena/software/sources/spack/var/spack/stage/caliper-2.0.1-abbxhgovgy4z7s3lw36rrbd4uvejyzfv/Caliper/include/caliper/common/util/spinlock.hpp:17:15: error: constexpr constructor never produces a constant expression [-Winvalid-constexpr]
    constexpr spinlock()
              ^
/Users/kumbhar/workarena/software/sources/spack/var/spack/stage/caliper-2.0.1-abbxhgovgy4z7s3lw36rrbd4uvejyzfv/Caliper/include/caliper/common/util/spinlock.hpp:18:11: note: non-constexpr constructor 'atomic_flag' cannot be used in a constant expression
        : m_lock { 0 }
          ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/atomic:1700:5: note: declared here
    atomic_flag(bool __b) _NOEXCEPT : __a_(__b) {} // EXTENSION
    ^
1 error generated.
make[2]: *** [src/services/CMakeFiles/caliper-services.dir/aggregate/Aggregate.cpp.o] Error 1
make[1]: *** [src/services/CMakeFiles/caliper-services.dir/all] Error 2
make: *** [all] Error 2
==> Error: ProcessError: Command exited with status 2:
    'make'
```

I am adding errors here instead of separate issue. I tried to also build master with AppleClang and got same issue.

With gcc@8.3.0 I am seeing below error (works with gcc@4.9.4):

```
± |llnl_develop → remotes/llnl/develop {17} ?:10 ✗| → spack install -v caliper@master %gcc ^python@2.7
==> cmake@3.5.2 : externally installed in /usr/local-fake
==> cmake@3.5.2 : already registered in DB
==> mpich@3.2 : externally installed in /usr/local/Cellar/open-mpi/4.0.0
==> mpich@3.2 : already registered in DB
==> python@2.7.10 : externally installed in /System/Library/Frameworks/Python.framework/Versions/2.7
==> python@2.7.10 : already registered in DB
==> Installing caliper
==> Searching for binary cache of caliper
==> Warning: No Spack mirrors are currently configured
==> No binary for caliper found: installing from source
==> Cloning git repository: https://github.com/LLNL/Caliper.git
==> No checksum needed when fetching with git
==> Already staged caliper-master-sf3xofge5vqrsapg2rqgibvrz2omfxu6 in /Users/kumbhar/workarena/software/sources/spack/var/spack/stage/caliper-master-sf3xofge5vqrsapg2rqgibvrz2omfxu6
==> No patches needed for caliper
==> Building caliper [CMakePackage]
==> Executing phase: 'cmake'
==> [2019-03-19-21:32:44.231834] 'cmake' '/Users/kumbhar/workarena/software/sources/spack/var/spack/stage/caliper-master-sf3xofge5vqrsapg2rqgibvrz2omfxu6/Caliper' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:PATH=/Users/kumbhar/workarena/software/sources/spack/opt/spack/darwin-mojave-x86_64/gcc-8.3.0/caliper-master-sf3xofge5vqrsapg2rqgibvrz2omfxu6' '-DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_FIND_FRAMEWORK:STRING=LAST' '-DCMAKE_FIND_APPBUNDLE:STRING=LAST' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=FALSE' '-DCMAKE_INSTALL_RPATH:STRING=/Users/kumbhar/workarena/software/sources/spack/opt/spack/darwin-mojave-x86_64/gcc-8.3.0/caliper-master-sf3xofge5vqrsapg2rqgibvrz2omfxu6/lib;/Users/kumbhar/workarena/software/sources/spack/opt/spack/darwin-mojave-x86_64/gcc-8.3.0/caliper-master-sf3xofge5vqrsapg2rqgibvrz2omfxu6/lib64;/usr/local/Cellar/open-mpi/4.0.0/lib' '-DCMAKE_PREFIX_PATH:STRING=/System/Library/Frameworks/Python.framework/Versions/2.7;/usr/local/Cellar/open-mpi/4.0.0;/usr/local-fake' '-DBUILD_TESTING=Off' '-DBUILD_DOCS=Off' '-DBUILD_SHARED_LIBS=On' '-DWITH_DYNINST=Off' '-DWITH_CALLPATH=Off' '-DWITH_GOTCHA=Off' '-DWITH_PAPI=Off' '-DWITH_LIBPFM=Off' '-DWITH_SOSFLOW=Off' '-DWITH_SAMPLER=Off' '-DWITH_MPI=On' '-DMPI_C_COMPILER=/usr/local/Cellar/open-mpi/4.0.0/bin/mpicc' '-DMPI_CXX_COMPILER=/usr/local/Cellar/open-mpi/4.0.0/bin/mpic++'
-- The C compiler identification is GNU 8.3.0
-- The CXX compiler identification is GNU 8.3.0
-- Checking whether C compiler has -isysroot
-- Checking whether C compiler has -isysroot - yes
-- Checking whether C compiler supports OSX deployment target flag
-- Checking whether C compiler supports OSX deployment target flag - yes
-- Check for working C compiler: /Users/kumbhar/workarena/software/sources/spack/lib/spack/env/gcc/gcc
-- Check for working C compiler: /Users/kumbhar/workarena/software/sources/spack/lib/spack/env/gcc/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Checking whether CXX compiler has -isysroot
-- Checking whether CXX compiler has -isysroot - yes
-- Checking whether CXX compiler supports OSX deployment target flag
-- Checking whether CXX compiler supports OSX deployment target flag - yes
-- Check for working CXX compiler: /Users/kumbhar/workarena/software/sources/spack/lib/spack/env/gcc/g++
-- Check for working CXX compiler: /Users/kumbhar/workarena/software/sources/spack/lib/spack/env/gcc/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - no
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - not found
CMake Error at /usr/local/Cellar/cmake/3.12.3/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find Threads (missing: Threads_FOUND)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.12.3/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.12.3/share/cmake/Modules/FindThreads.cmake:205 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:220 (find_package)


-- Configuring incomplete, errors occurred!
See also "/tmp/kumbhar/spack-stage/spack-stage-SpbKZu/Caliper/spack-build/CMakeFiles/CMakeOutput.log".
See also "/tmp/kumbhar/spack-stage/spack-stage-SpbKZu/Caliper/spack-build/CMakeFiles/CMakeError.log".
==> Error: ProcessError: Command exited with status 1:
    'cmake' '/Users/kumbhar/workarena/software/sources/spack/var/spack/stage/caliper-master-sf3xofge5vqrsapg2rqgibvrz2omfxu6/Caliper' '-G' 'Unix Makefiles' '-DCMAKE_INSTALL_PREFIX:PATH=/Users/kumbhar/workarena/software/sources/spack/opt/spack/darwin-mojave-x86_64/gcc-8.3.0/caliper-master-sf3xofge5vqrsapg2rqgibvrz2omfxu6' '-DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo' '-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON' '-DCMAKE_FIND_FRAMEWORK:STRING=LAST' '-DCMAKE_FIND_APPBUNDLE:STRING=LAST' '-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=FALSE' '-DCMAKE_INSTALL_RPATH:STRING=/Users/kumbhar/workarena/software/sources/spack/opt/spack/darwin-mojave-x86_64/gcc-8.3.0/caliper-master-sf3xofge5vqrsapg2rqgibvrz2omfxu6/lib;/Users/kumbhar/workarena/software/sources/spack/opt/spack/darwin-mojave-x86_64/gcc-8.3.0/caliper-master-sf3xofge5vqrsapg2rqgibvrz2omfxu6/lib64;/usr/local/Cellar/open-mpi/4.0.0/lib' '-DCMAKE_PREFIX_PATH:STRING=/System/Library/Frameworks/Python.framework/Versions/2.7;/usr/local/Cellar/open-mpi/4.0.0;/usr/local-fake' '-DBUILD_TESTING=Off' '-DBUILD_DOCS=Off' '-DBUILD_SHARED_LIBS=On' '-DWITH_DYNINST=Off' '-DWITH_CALLPATH=Off' '-DWITH_GOTCHA=Off' '-DWITH_PAPI=Off' '-DWITH_LIBPFM=Off' '-DWITH_SOSFLOW=Off' '-DWITH_SAMPLER=Off' '-DWITH_MPI=On' '-DMPI_C_COMPILER=/usr/local/Cellar/open-mpi/4.0.0/bin/mpicc' '-DMPI_CXX_COMPILER=/usr/local/Cellar/open-mpi/4.0.0/bin/mpic++'

1 error found in build log:
     29    -- Check if compiler accepts -pthread
     30    -- Check if compiler accepts -pthread - no
     31    -- Looking for pthread_create in pthreads
     32    -- Looking for pthread_create in pthreads - not found
     33    -- Looking for pthread_create in pthread
     34    -- Looking for pthread_create in pthread - not found
  >> 35    CMake Error at /usr/local/Cellar/cmake/3.12.3/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
     36      Could NOT find Threads (missing: Threads_FOUND)
     37    Call Stack (most recent call first):
     38      /usr/local/Cellar/cmake/3.12.3/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
     39      /usr/local/Cellar/cmake/3.12.3/share/cmake/Modules/FindThreads.cmake:205 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
     40      CMakeLists.txt:220 (find_package)
     41

See build log for details:
```

--- a/include/caliper/common/util/spinlock.hpp
+++ b/include/caliper/common/util/spinlock.hpp
@@ -10,12 +10,11 @@ namespace util
 {
     
 class spinlock {
-    std::atomic_flag m_lock;
+    std::atomic_flag m_lock = ATOMIC_FLAG_INIT;
 
 public:
 
-    constexpr spinlock()
-        : m_lock { 0 }
+    spinlock()
         { }
 
     void lock() {

--- a/src/caliper/Blackboard.h
+++ b/src/caliper/Blackboard.h
@@ -75,7 +75,7 @@ class Blackboard {
     
 public:
 
-    CONSTEXPR_UNLESS_PGI Blackboard()
+    Blackboard()
         : hashtable       {   },
           ref_toc         { 0 },
           ref_toctoc      { 0 },

